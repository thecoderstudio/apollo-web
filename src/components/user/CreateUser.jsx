import React from 'react';
import styled from 'styled-components';
import axios from 'axios';
import { Formik } from 'formik';
import { handleHTTPResponse } from '../../actions/error';
import { createUserSchema } from '../../validation/user';
import media from '../../util/media';
import Button from '../buttons/Button';
import OutlinedButton from '../buttons/OutlinedButton';
import Input from '../Input';
import Card from '../Card';

const Container = styled(Card)`
  position: absolute;
  left: 0;
  right: 0;
  top: 50%;
  margin-left: auto;
  margin-right: auto;
  min-height: 450px;
  width: 100%;
  max-width: 400px;
  transform: translateY(-50%);

  display: flex;
  flex-direction: column;
  justify-content: center;
  box-sizing:border-box;

  ${
    media.phone`
      padding: 16px;
      height: 100%;
    `
  }
`;

const Content = styled.div`
  height: 100%;
  max-height: 450px;
  display: grid;
  grid-template-rows: [title] 1fr [form ]4fr;
  align-items: center;
`;

const Title = styled.h2`
  grid-row: title;
  text-align: center;
`;

const Form = styled.form`
  height: 95%;
  grid-row: form;
  display: grid;
  grid-template-rows: auto;
  grid-row-gap: 40px;
`;

const Buttons = styled.div`
  display: flex;
  justify-content: space-around;

  &> button {
    min-width: 150px;
  }

  ${
    media.phone`
      justify-content: space-between;
      flex-direction: column-reverse;
    `
  }
`;

export default class CreateUser extends React.PureComponent {
  constructor(props) {
    super(props);
    this.createUser = this.createUser.bind(this);
    this.close = this.close.bind(this);
    this.state = {
      username: '',
      password: '',
      confirmPassword: ''
    };
  }

  createUser(credentials, { setErrors }) {
    axios.post(
      `${process.env.APOLLO_HTTP_URL}user`,
      credentials,
      { withCredentials: true }
    ).then(res => {
      this.close(true);
    })
    .catch(error => {
      handleHTTPResponse(error.response, true, true);
      if (error.response.status === 400) {
        setErrors({
          username: error.response.data.username.msg
        });
      }
    });
  }

  close(success=false) {
    this.props.onClose(success);
  }

  render() {
    return (
      <Container>
        <Content>
          <Title>Create new user</Title>
          <Formik
            initialValues={{ username: '', password: '', confirmPassword: '' }}
            validationSchema={createUserSchema}
            onSubmit={this.createUser}>
            {({ values, errors, handleChange, handleSubmit }) => (
              <Form onSubmit={handleSubmit}>
                <Input
                  name="username"
                  type="username"
                  placeholder="Username"
                  value={values.username}
                  error={errors.username}
                  onChange={handleChange} />
                <Input
                  name="password"
                  type="password"
                  placeholder="Password"
                  value={values.password}
                  error={errors.password}
                  onChange={handleChange}
                  autocomplete="new-password" />
                <Input
                  name="confirmPassword"
                  type="password"
                  placeholder="Confirm password"
                  value={values.confirmPassword}
                  error={errors.confirmPassword}
                  onChange={handleChange}
                  autocomplete="new-password" />
                <Buttons>
                  <OutlinedButton onClick={this.close}>Cancel</OutlinedButton>
                  <Button>Create user</Button>
                </Buttons>
              </Form>
            )}
          </Formik>
        </Content>
      </Container>
    );
  }
}
