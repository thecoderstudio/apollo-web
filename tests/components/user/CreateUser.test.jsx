import React from 'react';
import renderer from 'react-test-renderer';
import axios from 'axios';
import waitForExpect from 'wait-for-expect';
import { StatusCodes } from 'http-status-codes';
import CreateUser from '../../../src/components/user/CreateUser';

jest.mock('axios');

function getComponent(onClose) {
  return renderer.create(<CreateUser onClose={onClose} />);
}

function submitForm(root, username, password, confirmPassword) {
  const usernameInput = root.findByProps({type: 'username'});
  const passwordInput = root.findByProps({placeholder: 'Password'});
  const confirmPasswordInput = root.findByProps({placeholder: 'Confirm password'});
  const form = root.findByType('form');

  usernameInput.props.onChange({ currentTarget: {
    name: 'username',
    value: username
  }});
  passwordInput.props.onChange({ currentTarget: {
    name: 'password',
    value: password
  }});
  confirmPasswordInput.props.onChange({ currentTarget: {
    name: 'confirmPassword',
    value: confirmPassword
  }});
  form.props.onSubmit();
}

describe("create user", () => {
  beforeEach(() => {
    process.env = {
      APOLLO_HTTP_URL: 'http://localhost:1234/'
    };
  });

  it("renders correctly", () => {
    const tree = getComponent().toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("cancels", () => {
    const onClose = jest.fn();
    const component = getComponent(onClose);
    const root = component.root;

    root.findAllByType('button')[0].props.onClick();
    expect(onClose).toHaveBeenCalledWith(false);
  });

  it("validates input", async () => {
    const onClose = jest.fn();
    const component = getComponent(onClose);
    const root = component.root;

    submitForm(root, '', 'password', 'password');
    submitForm(root, 'test', 'pass', 'pass');
    submitForm(root, 'test', 'password', 'password2');

    await waitForExpect(() => {
      expect(axios.post).not.toHaveBeenCalled();
      expect(onClose).not.toHaveBeenCalledWith(true);
    });
  });

  it("posts successfully", async () => {
    const onClose = jest.fn();
    const component = getComponent(onClose);
    const root = component.root;

    axios.post.mockResolvedValue({
      status: StatusCodes.CREATED
    });

    root.findByType('form').props.onSubmit({ preventDefault: jest.fn() });
    submitForm(root, 'test', 'password', 'password');

    await waitForExpect(() => {
      expect(onClose).toHaveBeenCalledWith(true);
    });
  });

  it("handles unsuccessful post", async () => {
    const onClose = jest.fn();
    const component = getComponent(onClose);
    const root = component.root;

    axios.post.mockImplementationOnce(() => Promise.reject({
      response: {
        status: StatusCodes.BAD_REQUEST,
        statusText: "Bad request",
        data: {}
      }
    }));

    submitForm(root, 'test', 'password123', 'password123');

    axios.post.mockImplementationOnce(() => Promise.reject({
      response: {
        status: StatusCodes.INTERNAL_SERVER_ERROR,
        statusText: "Something went wrong",
        data: {}
      }
    }));

    submitForm(root, 'test', 'password123', 'password123');

    await waitForExpect(() => {
      expect(axios.post).toHaveBeenCalledTimes(2);
      expect(onClose).not.toHaveBeenCalledWith(true);
    });
  });
});
