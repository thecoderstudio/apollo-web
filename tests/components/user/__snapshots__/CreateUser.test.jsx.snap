// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create user renders correctly 1`] = `
.c9 {
  max-height: 50px;
  padding: 12px;
  border: none;
  border-radius: 5px;
  color: white;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 700;
  cursor: pointer;
}

.c9:hover {
  opacity: 0.90;
}

.c9:focus {
  outline: none;
}

.c9:active {
  opacity: 0.80;
}

.c9:disabled:hover {
  opacity: 1;
  cursor: auto;
}

.c10 {
  background-color: transparent;
  border: 1px solid;
}

.c6 {
  height: 80px;
}

.c7 {
  min-height: 50px;
  width: 100%;
  border: none;
  border-radius: 5px;
  font-family: 'Libre Franklin',sans-serif;
  font-size: 1rem;
  padding-left: 10px;
  opacity: 0.9;
  border: 1px solid transparent;
}

.c7:focus {
  outline: none border:1px solid;
}

.c1 {
  box-shadow: 0 10px 15px -3px rgba(0,0,0,0.1),0 4px 6px -2px rgba(0,0,0,0.05);
  padding: 40px;
  border-radius: 10px;
}

.c0 {
  display: inline;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c2 {
  position: absolute;
  left: 0;
  right: 0;
  top: 50%;
  margin-left: auto;
  margin-right: auto;
  min-height: 450px;
  width: 100%;
  max-width: 400px;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  box-sizing: border-box;
}

.c3 {
  height: 100%;
  max-height: 450px;
  display: grid;
  grid-template-rows: [title] 1fr [form ]4fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  grid-row: title;
  text-align: center;
}

.c5 {
  height: 95%;
  grid-row: form;
  display: grid;
  grid-template-rows: auto;
  grid-row-gap: 40px;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c8> button {
  min-width: 150px;
}

@media (max-width:47.9375em) {
  .c2 {
    padding: 16px;
    height: 100%;
  }
}

@media (max-width:47.9375em) {
  .c8 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
    -webkit-flex-direction: column-reverse;
    -ms-flex-direction: column-reverse;
    flex-direction: column-reverse;
  }
}

<div
  className="c0"
  onClick={[Function]}
>
  <div
    className="c1 c2"
  >
    <div
      className="c3"
    >
      <h2
        className="c4"
      >
        Create new user
      </h2>
      <form
        className="c5"
        onSubmit={[Function]}
      >
        <div
          className="c6"
        >
          <input
            className="c7"
            name="username"
            onChange={[Function]}
            placeholder="Username"
            type="username"
            value=""
          />
        </div>
        <div
          className="c6"
        >
          <input
            className="c7"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="c6"
        >
          <input
            className="c7"
            name="confirmPassword"
            onChange={[Function]}
            placeholder="Confirm password"
            type="password"
            value=""
          />
        </div>
        <div
          className="c8"
        >
          <button
            className="c9 c10"
            onClick={[Function]}
          >
            Cancel
          </button>
          <button
            className="c9"
          >
            Create user
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
